### PR TITLE
Add "Random" colormap for visualizing image segmentation maps

### PR DESCRIPTION
**Pull request by bmorris3**
### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/default/plugins/plot_options/plot_options.py [hunk 1]
```diff
@@ -1087,3 +1111,13 @@ def _is_image_viewer(viewer):
...
+
+    @observe('image_colormap_value')
+    def _random_cmap_limit_update(self, *args, **kwargs):
+        # if 'Random' colormap is used for visualizing image segmentation,
+        # ensure the stretch limits are the min and max, so all label colors
+        # are unique:
+        if self.image_colormap_value != 'Random':
+            return
+
+        self.stretch_preset.value = 100
```
Comment (kecnry) [inline jdaviz/configs/default/plugins/plot_options/plot_options.py:1123]: I think I would prefer this to just toggle an in-plugin warning.  I can even see how it would be useful to set non-min/max in order to "exclude" outliers, but I can see how that at least needs some sort of explanatory message and perhaps link to a section in the docs explaining this implementation and the nuances.

Comment (larrybradley) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:1123]: @kecnry For a segmentation image, the default here needs to be min/max.  Nothing else really makes sense.  Having a non-min/max would clip a lower and upper range of integer labels (which are generally at the bottom and top of the segmentation image) and thus different sources would get assigned the same colors.  There is no "outlier values" in a segmentation image.  Each source gets assigned a unique integer value. 

As an example of why non-min/max is very bad, the background (no detected sources) is assigned a label value of zero. If you were to assign say all values < 10 to the same color, then you couldn't even distinguish between background and sources, defeating the whole point.

This must use min/max.  An appropriate message can be shown stating the stretch levels have been changed to min/max.

Comment (kecnry) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:1123]: Ok, then can we detect that the image is a segmentation image and only show this as an option in that case?  I'm just worried about the more common case where an image is _not_ a segmentation image, the user selects "random" not knowing the implications, and things are changed automatically and perhaps they loose their carefully set stretch limits.

Comment (larrybradley) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:1123]: Segmentation images contain only integer data, but that's not a unique identifier.  Not sure of any way to automatically detect a segmentation image.

Comment (larrybradley) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:1123]: Segmentation images often contain consecutive integer values (0 to max label), but that's definitely not a requirement either.

Comment (kecnry) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:1123]: Maybe this would be better as a separate plugin (or button within plot options that explicitly states all the things that will be set - similar to the RGB defaults button) then?  Otherwise, I think we're adding confusion for the default case in order to support a case that isn't relevant for most situations.

**Comment by pllim** [comment]
tl;dr -- How much of this becomes irrelevant (i.e., we have to remove/refactor) if we end up with a separate "DQ" plugin?

**Comment by larrybradley** [comment]
DQ and segmentation arrays aren't the same thing (although both do contain integer-only data).

**Comment by bmorris3** [comment]
@pllim We resolved earlier in the series of tickets that image segmentation is not DQ, and we shouldn't conflate them.

**Comment by pllim** [comment]
"Random" in the dropdown sounds too cryptic to me. What about "Segmentation"?

File: CHANGES.rst [hunk 2]
```diff
@@ -24,6 +24,8 @@ Cubeviz
 Imviz
 ^^^^^
 
+- Add "Random" colormap for visualizing image segmentation maps [#2671]
```
Comment (pllim) [inline CHANGES.rst:27]:
```suggestion
- Add "Random" colormap for visualizing image segmentation maps. [#2671]
```

**Review by pllim — COMMENTED** [review]
The doc here needs to be expanded as well: https://github.com/spacetelescope/jdaviz/blob/main/docs/imviz/displayimages.rst#colormap

**Comment by bmorris3** [comment]
I've added a warning when the user selects the "Random" colormap, indicating which other preferences have been changed.

Screen Shot 2024-01-24 at 14 15 09

File: jdaviz/configs/default/plugins/plot_options/plot_options.py (hunk 1 above)
Comment (kecnry) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:1129]: > One plugin which controls plot options is Plot Options.

I know, and I agree it makes most sense here, but then it needs to act in a way that makes sense for any input data in which the choice is available.  If you're someone who isn't even familiar with segmentation maps and load an image and stumble on this, it will be very jarring and confusing.

> If you're intentionally using this button for its intended purpose, this is adding clicks that you will always have to go through.

Yes, that is the downside, but I think it makes sense to add the extra clicks to the specific use-case rather than the general use-case.  That said, the default stretch function is already linear, so by default this will just be two clicks, not three.  If the parser could detect segmentation maps, then we could even have the percentile default for those layers to min/max from load, and then this only requires extra clicks if the user overrides anything.

Another benefit to this design is if the user changes to "Image Segmentation" (or "Random") and _then_ changes the stretch inputs, the warning will re-appear.  The current implementation has no way to catch that case unless it disables those inputs entirely.

> Would your concerns be assuaged if we named the cmap "Image Segmentation"?

It is slightly more specific than "Random", but I still don't think it provides the necessary context and explanation for someone who does not have this use-case in their mind at all when looking through the list of options for colormap.

**Comment by kecnry** [comment]
> I've added a warning when the user selects the "Random" colormap, indicating which other preferences have been changed.

I still prefer the opposite approach for reasons explained above, but could let it go if there are revert buttons.  Is this warning exposed from the API as well?

**Comment by rosteen** [comment]
Changing the contrast and bias can also mess up the uniqueness of the color map - do we want to either disable those in Random colormap mode, or handle them the same way as stretch preset and function?


https://github.com/spacetelescope/jdaviz/assets/39831871/01164fd5-fb40-42a5-97ff-d5f5e16b3762

**Comment by bmorris3** [comment]
@rosteen – now enforcing sensible bias and contrast presets in 
a50aca4.

**Review by kecnry — APPROVED** [review]
LGTM now - thanks for the catch on the contrast/bias @rosteen and the fix @bmorris3!

**Review by rosteen — APPROVED** [review]
LGTM as well now too, thanks.
